src/Tools/Source/CompilerGeneratorTools/Source/CSharpSyntaxGenerator/SourceGenerator.cs

```diff
@@ -123,13 +122,40 @@ void addResult(Action<TextWriter> writeFunction, string hintName)
                 {
                     writeFunction(textWriter);
                 }
+                return stringBuilder;
+            }
 
+            static void addResult(SourceProductionContext context, StringBuilder stringBuilder, string hintName)
+            {
                 // And create a SourceText from the StringBuilder, once again avoiding allocating a single massive string
-                var sourceText = SourceText.From(new StringBuilderReader(stringBuilder), stringBuilder.Length, encoding: Encoding.UTF8);
-                sourcesBuilder.Add((hintName, sourceText));
+                context.AddSource(hintName, SourceText.From(new StringBuilderReader(stringBuilder), stringBuilder.Length, encoding: Encoding.UTF8));
             }
         }
 
+        private record Result();
+
+        private record SuccessResult(StringBuilder Main, StringBuilder Internal, StringBuilder Syntax) : Result;
+
+        private record DiagnosticResult(Diagnostic Diagnostic) : Result;
+
+        private record SyntaxInfo(AdditionalText? File, SourceText? Text = null);
+
+        private sealed class ResultComparer : IEqualityComparer<Result>
+        {
+            public bool Equals(Result x, Result y)
+            {
+                if (x is SuccessResult sx && y is SuccessResult sy)
+                {
+                    return sx.Main.Equals(sy.Main)
+                        && sx.Internal.Equals(sy.Internal)
+                        && sx.Syntax.Equals(sy.Syntax);
+                }
+                return x.Equals(y);
```

Comment:
Why the `if` is needed? I'm expecting Equals to work since these should have a synthesized Equals.

Reply:
I agree, what is the point of the ``if``?